### DISCUSSION TITLE
Reading Coil

### DISCUSSION THREAD

**Discussion by Elmatbaisaad** [aldas/modbus-tcp-client#111] [Q&A]
Hello everyone i m working on a project where i must read coils and registers i succed in reading registers but not in coils .
I m trying this code but its not working to exlain what i want :
- i want to give and addresse and in result i want the value of the address.
![coil](https://user-images.githubusercontent.com/101296742/163959322-c96233a2-75ed-4225-80d7-86bfbd7579a6.png)

**Comment by aldas** [comment]
What does it say? Is there an error?

This is how I test:

1. I'll download modbus server from https://www.modbusdriver.com/diagslave.html
2. start it with `./diagslave -m tcp -p 5022`
3. Use this php code to test (file named `coils_builder.php`)
```php
<?php

require __DIR__ . '/../vendor/autoload.php';

use ModbusTcpClient\Composer\Read\ReadCoilsBuilder;
use ModbusTcpClient\Network\NonBlockingClient;

$fc3 = ReadCoilsBuilder::newReadCoils('tcp://127.0.0.1:5022')
    ->coil(0, "my_coil")
    ->build();

$responseContainer = (new NonBlockingClient(['readTimeoutSec' => 0.2]))->sendRequests($fc3);
var_dump($responseContainer->getData()); // array of assoc. arrays (keyed by address name)
var_dump($responseContainer->getErrors());
```
4. Execute the script
```bash
x@x:~/code/modbus-tcp-client/examples$ php -v
PHP 8.1.2 (cli) (built: Jan 24 2022 10:42:51) (NTS)
Copyright (c) The PHP Group
Zend Engine v4.1.2, Copyright (c) Zend Technologies
    with Zend OPcache v8.1.2, Copyright (c), by Zend Technologies
    with Xdebug v3.1.2, Copyright (c) 2002-2021, by Derick Rethans

x@x:~/code/modbus-tcp-client/examples$ php coils_builder.php 
/home/x/code/modbus-tcp-client/examples/coils_builder.php:13:
array(1) {
  'my_coil' =>
  bool(false)
}
/home/x/code/modbus-tcp-client/examples/coils_builder.php:14:
array(0) {
}
```
6. Diagslave output is as:
```bash
x@x:~/code$ ./diagslave -m tcp -p 5022
diagslave 3.1 - FieldTalk(tm) Modbus(R) Diagnostic Slave Simulator
Copyright (c) 2002-2018 proconX Pty Ltd
Visit https://www.modbusdriver.com for Modbus libraries and tools.

Protocol configuration: MODBUS/TCP
Slave configuration: address = -1, master activity t/o = 3.00
IP configuration: port = 5022, connection t/o = 60.00

Server started up successfully.
Listening to network (Ctrl-C to stop)
............
validateMasterIpAddr: accepting connection from 127.0.0.1
Slave   0: readCoils from 1, 1 references
....................................................
```

**Comment by Elmatbaisaad** [comment]
i got this error 
![err](https://user-images.githubusercontent.com/101296742/163971697-52784fcf-40d7-4ab6-bf36-31d484a80be5.png)

**Reply by aldas** [reply]
what version of PHP are you using?

**Reply by Elmatbaisaad** [reply]
I am using php 8.0

**Reply by aldas** [reply]
in your first example the line `require __DIR__ . '/../vendor/autoload.php';` is after `use ...` lines. maybe if you move `require` to be before `use`.

**Reply by Elmatbaisaad** [reply]
I've tried it but i got the same error is my code working for you ??

**Reply by aldas** [reply]
I can not test code that is pasted as image.

**Comment by Elmatbaisaad** [comment]
Here is my code, I really don't know where this come from
[CoilRead.txt](https://github.com/aldas/modbus-tcp-client/files/8511776/CoilRead.txt)

**Reply by aldas** [reply]
This code works for me.  I have changed the url to `tcp://127.0.0.1:5022` just because this is where modbus server is running for me but the output with PHP 8.0.18  is:

```bash
x@x:~/code/modbus-tcp-client$ docker run --rm --network="host" -v $(pwd):/code php:8.0.18-cli-alpine3.15 php /code/examples/read_coils.php
array(1) {
  ["my_coil"]=>
  bool(false)
}
array(0) {
}
```

**Reply by Elmatbaisaad** [reply]
i keep getting the error :/

**Reply by aldas** [reply]
which version of this library are you using?

**Reply by Elmatbaisaad** [reply]
i am using 3.0

**Reply by aldas** [reply]
Then I really do not know `ReadCoilsBuilder` is not implementing/extending anything that would require it to implement those 4 methods. I really do not why you are getting this error.

You could try to delete that library folder and reinstall with composer.

**Reply by Elmatbaisaad** [reply]
i reinstalled the library that's work but i have no value

**Reply by Elmatbaisaad** [reply]
i have a json file where i put adresses and values for example i put for coils 
"adresse":9
value:true // because its a bit
but it returns nothing
